packages/tenant-enrollment-nodejs/src/pluginImplementation.ts

```diff
@@ -0,0 +1,192 @@
+        }
+      } else if (userIdentificationDetail.type === "thirdParty") {
+        canJoin = this.isApprovedIdPProvider(tenantId, userIdentificationDetail.thirdPartyId);
+        if (!canJoin) {
+          reason = this.getMessageForNoSignUpReason(NotAllowedToSignUpReason.IDP_NOT_ALLOWED);
+        }
+      } else if (userIdentificationDetail.type === "phoneNumber") {
+        // We don't really have a way to check anything for phones so we can
+        // allow signup.
+        return {
+          canJoin: true,
+        };
+      }
+
+      return {
+        canJoin,
+        reason,
+      };
+    },
+    handleTenantJoiningApproval: async function (
+      user: User,
+      tenantId: string,
+      associateLoginMethodDef: AssociateAllLoginMethodsOfUserWithTenant,
+      sendEmail: SendPluginEmail,
+      appUrl: string,
+      userContext: UserContext,
+      assignRoleToUserInTenant: AssignRoleToUserInTenant,
+    ) {
+      /**
+       * Handle the tenant joining functionality for the user.
+       *
+       * If the tenant requires approval, we will add a request for the
+       * user.
+       * If the tenant doesn't require approval, we will add them as a member
+       * right away.
+       *
+       * @param user - The user to handle the tenant joining for
+       * @param tenantId - The id of the tenant to handle the tenant joining for
+       * @param associateLoginMethodDef - The function to associate the login methods of the user with the tenant
+       */
+      // Skip this for the public tenant
+      if (tenantId === "public") {
+        return {
+          wasAddedToTenant: true,
+          reason: undefined,
+        };
+      }
+
+      // If the tenant doesn't require approval, add the user as a member
+      // and return.
+      if (!this.doesTenantRequireApproval(tenantId)) {
+        await assignRoleToUserInTenant(tenantId, user.id, ROLES.TENANT_MEMBER);
+        return {
+          wasAddedToTenant: true,
+        };
+      }
+
+      // We don't need to do anything in particular except notifying
+      // the tenant admins about the new user request being added.
+      await this.sendTenantJoiningRequestEmail(tenantId, user, appUrl, sendEmail, userContext);
+
+      return {
+        wasAddedToTenant: false,
+        reason: "REQUIRES_APPROVAL",
+      };
+    },
+    isTenantInviteOnly: function (tenantId) {
+      return config.inviteOnlyTenants?.includes(tenantId) ?? false;
+    },
+    doesTenantRequireApproval: function (tenantId) {
+      return config.requiresApprovalTenants?.includes(tenantId) ?? false;
+    },
+    isApprovedIdPProvider: function (thirdPartyId) {
+      return thirdPartyId.startsWith("boxy-saml");
+    },
+    isMatchingEmailDomain: function (tenantId, email) {
+      const emailDomain = email.split("@");
+      if (emailDomain.length !== 2) {
+        return false;
+      }
+
+      const parsedTenantId = config.emailDomainToTenantIdMap[emailDomain[1]!.toLowerCase()];
```

Comment:
Would this make sense?
```suggestion
      const parsedTenantId = config.emailDomainToTenantIdMap[emailDomain[1]!.toLowerCase()] ?? emailDomain[1]!.toLowerCase();
```

Reply:
Yeah, I think we can do this.